src/core/mempool.rs

```diff
@@ -750,6 +756,113 @@ impl MemPoolTxInfo {
     }
 }
 
+/// Used to locally cache nonces to avoid repeatedly looking them up in the nonce.
+struct NonceCache {
+    cache: HashMap<StacksAddress, u64>,
+    size: usize,
+}
+
+impl NonceCache {
+    const MAX_SIZE: usize = 1024 * 1024;
+
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+            size: 0,
+        }
+    }
+
+    fn get<C>(&mut self, address: &StacksAddress, clarity_tx: &mut C) -> u64
+    where
+        C: ClarityConnection,
+    {
+        match self.cache.get(address) {
+            Some(nonce) => *nonce,
+            None => {
+                let nonce = StacksChainState::get_nonce(clarity_tx, &address.clone().into());
+                // Simple size cap to the cache -- once it's full, all nonces
+                // will be looked up every time. This is bad for performance
+                // but is unlikely to occur due to the typical number of
+                // transactions processed before filling a block.
+                if self.size < Self::MAX_SIZE {
+                    self.cache.insert(address.clone(), nonce);
+                    self.size += 1;
+                }
+                nonce
+            }
+        }
+    }
+
+    fn increment(&mut self, address: StacksAddress) {
+        let nonce = self.cache.entry(address).or_insert(0);
+        *nonce += 1;
+    }
+}
+
+struct CandidateCache {
```

Comment:
Can you add a docstring to this struct?  Thanks!

Reply:
Added in 27ee2f88c